### ISSUE TITLE
NVFP4 performance regression in TE main branch

### ISSUE THREAD

**Issue by cael-ling** [NVIDIA/TransformerEngine#2558]
**Describe the bug**

TransformerEngine [PR-2411](https://github.com/NVIDIA/TransformerEngine/pull/2411) introduced optimizations to improve NVFP4 computation performance. Previously, using an image built with this PR delivered significant performance gains during the pre-training of the Moonlight-16B model. This PR was merged into the TE main branch two weeks ago. However, when building an image from the main branch and running the same NVFP4 training configuration, the throughput per GPU dropped dramatically and became highly unstable.

**Steps/Code to reproduce bug**

To reproduce the issue, perform the following comparative experiments:

Step1: Build two Slurm images:
Refer to the script at [link](https://gitlab-master.nvidia.com/caell/te_issues/-/blob/main/20260105/build_te_image_0104_main_c988548.sh?ref_type=heads) to build two Slurm images, one image is based on PR-2411, and another image is based on TE main (commit id c988548f72bbc271fe2ab7bad1046b91b577aa29）

Step2: Launch a training job:
Refer to the script at [link](https://gitlab-master.nvidia.com/caell/te_issues/-/blob/main/20260105/eval_nvfp4_train_moonlight.sh?ref_type=heads) to launch a training job. It doesn’t necessarily have to be the Moonlight model, other models are also likely to experience similar performance issues.

**Expected behavior**

The training performance based on these two images are expected to be similar, but in reality, the performance gap is huge (570 TFLOPS VS 240-400 TFLOPS), see [log1](https://gitlab-master.nvidia.com/caell/te_issues/-/blob/main/20260105/result1.log?ref_type=heads),[log2](https://gitlab-master.nvidia.com/caell/te_issues/-/blob/main/20260105/result2.log?ref_type=heads), which makes NVFP4 training solution unusable. 

**Environment overview (please complete the following information)**

 - Environment location: Slurm cluster
 - Method of Transformer Engine install: please refer to [link](https://gitlab-master.nvidia.com/caell/te_issues/-/blob/main/20260105/build_te_image_0104_main_c988548.sh?ref_type=heads) 
 - If method of install is [Docker], provide `docker pull` & `docker run` commands used
please refer to [link](https://gitlab-master.nvidia.com/caell/te_issues/-/blob/main/20260105/eval_nvfp4_train_moonlight.sh?ref_type=heads)

**Comment by zhongbozhu** [comment]
Please double check padding & NVTE_USE_FAST_MATH toggle.

**Comment by zhongbozhu** [comment]
Here is the expected kernel trace for 128 multiple padding: 

<img width="1490" height="562" alt="Image" src="https://github.com/user-attachments/assets/362ed344-4a31-45ca-8110-ed55014f4d16" />

**Comment by zhongbozhu** [comment]
For 64 padding, we lose the row-col padding, so three kernels in quantization instead of two. 

<img width="1943" height="634" alt="Image" src="https://github.com/user-attachments/assets/7bcf76e9-5b77-4c34-b484-2d9a45e1a908" />

Without 64 padding, no group kernel will be performed. 

Padding overhead also a big factor without router padding.

**Comment by cael-ling** [comment]
> Please double check padding & NVTE_USE_FAST_MATH toggle.

export NVTE_USE_FAST_MATH=1
--enable-experimental 
--moe-router-padding-for-quantization

No changes after enabling these toggles

**Comment by cael-ling** [comment]
> Here is the expected kernel trace for 128 multiple padding:
> 
> <img alt="Image" width="1490" height="562" src="https://private-user-images.githubusercontent.com/42691305/532085826-362ed344-4a31-45ca-8110-ed55014f4d16.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc2Njg3NjksIm5iZiI6MTc2NzY2ODQ2OSwicGF0aCI6Ii80MjY5MTMwNS81MzIwODU4MjYtMzYyZWQzNDQtNGEzMS00NWNhLTgxMTAtZWQ1NTAxNGY0ZDE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA2VDAzMDEwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzYmUwMDZmOWNlOWY0NzBjMTk1MGQ4MjkwZjE5OTRhYWUzOTQ1MmNiZDY0MmVmMjg2MjY1NDFjOWExNmVlNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.k_lrtP0tkaq_1rdMpjYtH4dBBawbiB6xQ3K3SGTb79k">

I can see two kernels: group_amax and row_col_rht_gemm when use the image built from pr-2411. However, in the main-branch build, the timeline shows a long idle period inside groupedlinear, which appears to be caused by cudaMallocAsync, and this is severely impacting performance.

<img width="1609" height="554" alt="Image" src="https://github.com/user-attachments/assets/6e6459b7-46f7-4986-809c-529fc5b61c74" />


Long idle in the timeline using main branch build image to run the train:
<img width="1354" height="480" alt="Image" src="https://github.com/user-attachments/assets/1bd8df50-6bdd-411b-9ca8-57939c9ca034" />

**Comment by zhongbozhu** [comment]
You are right, this is a bug...

We shouldn't allocate this workspace for the atomic tile scheduler but instead making it a torch tensor that's always live. cc @timmoon10. 

We didn't have this problem because we had a clc based scheduler for this kernel which doesn't need a workspace buffer. However clc scheduler assumes knowing problem shape on host, which won't be cuda graph compatible in the future, so we removed it. 

<img width="3794" height="1325" alt="Image" src="https://github.com/user-attachments/assets/acb79410-e3ae-4f1f-b908-c05ad39172a8" />

**Comment by zhongbozhu** [comment]
bug fix https://github.com/NVIDIA/TransformerEngine/pull/2564

<img width="3788" height="1523" alt="Image" src="https://github.com/user-attachments/assets/233f89e1-8bd1-4e8f-a228-9642b04c4b4f" />

**Comment by zhongbozhu** [comment]
Will close this issue since it's resolved. @cael-ling can you confirm?

**Comment by cael-ling** [comment]
> Will close this issue since it's resolved. [@cael-ling](https://github.com/cael-ling) can you confirm?

Just tested it, and the issue is fixed now, please close the issue, nice work.
